### DISCUSSION TITLE
The subclass in Callable

### DISCUSSION THREAD

**Discussion by HouLingLXH** [python/typing#1133] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/30954632/163298365-cfc134dd-cdf9-4216-aa6a-f992680c9166.png)
![image](https://user-images.githubusercontent.com/30954632/163298427-076bc186-cf35-4d38-8477-f7543ba4fe16.png)

![image](https://user-images.githubusercontent.com/30954632/163298378-fe109c86-e1cb-4d62-b2bc-214cb1a66a02.png)

EventTest is the subclass of EventBase, how to fixed it?

**Comment by carljm** [comment, marked as answer]
You asked for a callable that can take any EventBase as an argument, you provided one that only accepts an EventTest. This is a type error because the thing you provided is "less capable" than the expected type. In type system terms, a callable type is "contravariant" in its accepted arguments: if you accept only a subtype of the arguments of another callable type, that makes you a supertype of it -- the relationship is reversed. 

In this case, you either need to provide a callback that can accept any EventBase, or you might want to use a typevar (eg `EventBaseT = TypeVar("EventBaseT", bound=EventBase)` to replace `EventBase` in the type annotations.

**Reply by carljm** [reply]
Yes, EventTest is more capable than EventBase (which is just another way of saying EventTest is a subtype of EventBase.) But for callable types this relationship is reversed for their arguments: a function that can only accept EventTest is less capable than a function that can accept any EventBase. Clearly the latter function can also accept an EventTest, so it is at least as capable.

If you are expecting a callback function that can take any EventBase, you would be correct to call the callback with an EventFoo. Now if someone passes you a callback that can accept only EventTest, your call with an EventFoo would be wrong! This is why it is a type error to pass a function accepting only EventTest where a function accepting any EventBase is required. 

In your specific case, if you are registering a listener for EventTest, then you know the callback will only be called with EventTest objects. This is why a type variable is the right approach for your situation: the type variable means the type of event you are listening for and the type of event your callback accepts must always be the same.

**Reply by HouLingLXH** [reply]
Thanks.
eg.
    TEventBase = TypeVar("TEventBase", bound=EventBase)
    def add_listener(self, event_class: TEventBase, want_do: Callable[[TEventBase], None]) -> Listener:
       ...

    def call_testA(self, data:AEvent)
       ...
    def call_back(self, data:BEvent)
      ...
    


I have two subclass (AEvent、 BEvent) of EventBase, if :

    add_listener(AEvent,call_back)

It is wrong in work but no hint in editor, becase I want to data of call_back must be AEvent when the first parm is AEvent.

  add_listener(AEvent,call_testA) √
  add_listener(BEvent,call_back) √
  add_listener(BEvent,call_testA) ×

How to hint it in function?
Thanks.

**Reply by TeamSpen210** [reply]
You want `event_class: type[TEventBase]` to indicate it requires the type itself not a subclass.

**Reply by HouLingLXH** [reply]
I want to rule the type of  param in Callable by the type of event_class , rather than only is the subclass of EventBase.
    add_listener(AEvent,call_testA) √
    add_listener(BEvent,call_back) √
    add_listener(BEvent,call_testA) ×

**Reply by Akuli** [reply]
As the answer says, you can use a TypeVar:

```python3
from typing import TypeVar, Callable

class EventBase:
    pass

class EventA(EventBase):
    pass

class EventB(EventBase):
    pass

T = TypeVar("T", bound=EventBase)

def add_listener(event_class: type[T], want_do: Callable[[T], None]) -> None:
    pass

def foo(data: EventA) -> None:
    pass

add_listener(EventA, foo)  # ok
add_listener(EventB, foo)  # error: Argument 2 to "add_listener" has incompatible type "Callable[[EventA], None]"; expected "Callable[[EventB], None]"
```

**Reply by HouLingLXH** [reply]
Why not error in my pycharm? 
![image](https://user-images.githubusercontent.com/30954632/163341760-03741f86-d2d6-45c0-944f-52ebaf84596c.png)

**Reply by carljm** [reply]
I don't use Pycharm so not sure. It's possible that pycharm is focused more on autocomplete etc and not so much on type errors and is more permissive than dedicated type checkers like mypy or pyre or pyright.
